### PR TITLE
Update hello-books examples in copilot lessons

### PR DESCRIPTION
**Pull request by kelsey-steven-ada**
[Asana Task](https://app.asana.com/0/1205655776549324/1206697443678577/f)

Updates to use newer Flask/SQLAlchemy syntax in code examples across the working with Copilot lessons. Replaces screenshots, code examples, and updates content to support the code generated while reworking this lesson. Includes some updates for changes to the Copilot UI since first writing, especially around the test generation UI.

### CODE REVIEW

File: working-w-copilot/copilot-in-projects-new-code.md [hunk 1]
```diff
@@ -40,7 +40,7 @@ We could use a comment to try generating our `from_dict` class method, but inste
 When we submit the prompt, we are likely to receive a response similar to the following:
 
 ![VS Code showing the Copilot inline chat with the prompt above entered and a function suggestion below](assets/copilot-in-projects/book-from-dict-initial-suggestion.png)  
-*Fig. Our prompt entered in the Copilot inline chat with a suggestion for the `from_dict` function from Copilot in a temporary window below*
+*Fig. Our prompt entered in the Copilot inline chat with a suggestion for the `from_dict` function from Copilot in a temporary window below ([Full size image](assets/copilot-in-projects/book-from-dict-initial-suggestion.png))*
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-new-code.md:60]: The current screen shot shows the change from the chat in side-by-side mode, but I'm seeing it directly inline. I tried toggling the source control diff setting and that didn't seem to have an effect, and the only copilot setting I found `inlineChat.mode` could show side-by-side, but not directly in the editor as in the screen shot. Does yours still appear that way, or is this a change in the preview behavior?

Screenshot 2024-11-20 at 15 45 08

File: working-w-copilot/copilot-in-projects-new-code.md [hunk 2]
```diff
@@ -88,59 +88,45 @@ These functions aren't very long, but it's still a good idea to test them as a b
 In our `book.py` file, highlight the whole text, bring up the inline Copilot chat with `⌘I` (`CMD + i`), then type in the shortcut `/tests`.
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-new-code.md:105]: When I did this, since I hadn't setup tests previously, copilot asked if it could start that flow. I was able to pick pytest, but at this point, the repo doesn't have a tests directory. I was able to escape out at the directory prompt, and then the /tests command worked, but there was some friction at the start that we might want to call out.

File: working-w-copilot/copilot-in-projects-new-code.md [hunk 3]
```diff
@@ -88,59 +88,45 @@ These functions aren't very long, but it's still a good idea to test them as a b
...
-
-The "Create" button will generate the new test file in the current folder, in our case, in the same folder as our `Book` model. This isn't really what we want, but we can click the arrow next to the "Create" button to see a drop down with a "Create As" option. This will let us change the default name of the file and choose where it gets saved so that we can keep all of our test files together, separate from our source code. If we don't use "Create As", we can drag and drop test files from where they were created to the test folder at any later time.
-
-<br />
-
-![Close up of the Copilot window for previewing tests showing the options drop down from the "Create" button](assets/copilot-in-projects/book-slash-tests-create-as.png)  
-*Fig. "Create As" in the drop down next to the "Create" button*
-
-### !end-callout
+If we feel like the tests presented are a good starting place, we can click "Accept" in the copilot chat to dismiss the chat window, then use `⌘S` (`CMD + S`) to save the new test file. We should see a pop up that allows us to choose the folder where we want to save the test file.
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-new-code.md]: Since the tests folder wasn't in this branch, it's not obvious where to save this file. To mirror our general setup, I had to create the tests directory and come up with the test file name (I looked in a future branch to recall what structure we used). I then also had to remember to add the `__init__.py` file myself. The import instruction is also written assuming the test file is in the same folder as the model, but I had to update that since I saved it to our more typical structure. Could we spell out more explicitly what we expect folks to do with the saving so they can focus more on the copilot aspect?

File: working-w-copilot/copilot-in-projects-new-code.md [hunk 4]
```diff
@@ -158,9 +144,9 @@ When we use Copilot to create tests, we should always evaluate if the data is be
...
+We have tests to ensure that for a given book model, the dictionary that `to_dict` creates contains the right data, and that a book created by `from_dict` stores the title and description from the input dictionary into the right properties. Our last test ensures that if the `"title"` key is missing from the input dictionary, we raise an error. But what if the `"description"` key is missing instead?
 
-To make our test suite as complete as possible, we can update the name of our last test to `test_book_from_dict_missing_description` and add one more test `test_book_from_dict_missing_title` to make sure that we have this edge case covered. We can write the additional test by hand, but we can also try highlighting the last test and asking Copilot to do that work for us!
+To make our test suite as complete as possible, we can add one more test `test_book_from_dict_missing_description` to make sure that we have this edge case covered. We can write the additional test by hand, but we can also try highlighting the last test and asking Copilot to do that work for us!
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-new-code.md:168]: Copilot generated this case for me from the outset. 😅

File: working-w-copilot/copilot-in-projects-new-code.md [hunk 5]
```diff
@@ -231,142 +208,123 @@ Now that we've reminded ourselves of what the classes need to do, we can use the
...
 >         return {
 >             'id': self.id,
 >             'title': self.title,
->             'description': self.description
->         }
->
+>             'description': self.description,
+>             'author': self.author.to_dict() if self.author else None,
+>             'genres': [genre.to_dict() for genre in self.genres]
+>         }
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-new-code.md:332]: Copilot only suggested adding the author_id to the dict output for me, rather than adding the dictionaries of the related records.

File: working-w-copilot/copilot-in-projects-new-code.md [hunk 6]
```diff
@@ -386,61 +344,65 @@ We could make the changes manually after moving the code to a file, but we can a
 
 <details>
   <summary>
-    We will leave marking properties as required as a personal exercise, but feel free to try out asking Copilot to help with the changes to the <code>Book</code> and <code>Author</code> <code>to_dict</code> functions. When you're done, expand this section to see the prompt we used and what Copilot suggested.
+    Feel free to try out asking Copilot to help with the changes to the <code>Book</code> and <code>Author</code> <code>to_dict</code> functions. When you're done, expand this section to see the prompt we used and what Copilot suggested.
   </summary>
 
   **Updated Prompt:**
-  > Please update the Book to_dict function to include the name of the author and a list of the names of the genres associated in the return dictionary, only if those pieces of data exist. Please update the Author to_dict function to include the title of each item in the list books.
+  > Please update the Book to_dict function to include only the name of the author and a list of the names of the genres associated in the return dictionary, if those pieces of data exist. Please update the Author to_dict function to include only the title of each item in the list books.
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-new-code.md:370]: Even though author and genre were absent from my copilot-generated to_dict, this prompt still got it to the same place.

File: working-w-copilot/copilot-in-projects-refactoring.md [hunk 7]
```diff
@@ -12,13 +12,13 @@ Our goals for this lesson are to:
 
 Writing new code with Copilot is great, but a significant portion of software development involves refactoring and updating existing code. We got a little taste of updating code with the `Book` class, but we're going to revisit a scenario that we've seen before—refactoring the `validate_book` function into a more generic function that we can use for any model.
 
-For this section of the lesson, we're going to check out a different `hello-books` branch, [`07b-from-dict-refactor`](https://github.com/AdaGold/hello-books-api/tree/07b-from-dict-refactor), as the starting point.
+For this section of the lesson, we're going to check out a different `hello-books` branch, [`07b-to-dict-refactor`](https://github.com/AdaGold/hello-books-api/tree/07b-to-dict-refactor), as the starting point.
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-refactoring.md:31]: Oh no. After we undo the refactor order swap at sme point in the future, this will will need to go back to what was there before. 😅

I added a subtask to that swap back ticket to revist this: https://app.asana.com/0/1201700438643002/1208793757696414/f

File: working-w-copilot/copilot-in-projects-refactoring.md [hunk 8]
```diff
@@ -76,36 +76,66 @@ Searching our project, we find 4 dependencies:
...
+def test_validate_book_invalid_id(client, two_saved_books):
+    # Act & Assert
+    with pytest.raises(BadRequest):
+        validate_book("cat")
+```
+
+These tests are getting closer to what we need, but still won't work for us. The main issues are that:
+- The `validate_book` import path at the top of the file is slightly wrong
+- The tests require `pytest` to use `with pytest.raises`
+- The symbols `NotFound` and `BadRequest` used in the edge case tests do not exist
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-refactoring.md:150]: Mine came up with ValueError checks. Still wrong, but in fun and exciting ways!
```py
    def test_validate_book_invalid_id(client, two_saved_books):
        # Act & Assert
        with pytest.raises(ValueError) as excinfo:
            validate_book("cat")
        assert str(excinfo.value) == "book cat invalid"

    def test_validate_book_missing_record(client, two_saved_books):
        # Act & Assert
        with pytest.raises(ValueError) as excinfo:
            validate_book(3)
        assert str(excinfo.value) == "book 3 not found"
```

File: working-w-copilot/copilot-in-projects-refactoring.md [hunk 9]
```diff
@@ -76,36 +76,66 @@ Searching our project, we find 4 dependencies:
...
+- The status code isn't being checked on the edge cases where the request should have been aborted
+- The `client` fixture is not required on any of the tests since we aren't making a request to an endpoint
+
+Fixing the imports is a couple lines of code, but for our edge cases, we may need to do some research to be sure of the error we expect to be raised by our functions and ensure that our tests are looking for that particular error to be raised. In this case, we know from the Building an API series that the `validate_book` function will raise an `HTTPException` that we need to import from `werkzeug.exceptions` to access in our tests.
 
 ### !callout-info
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-refactoring.md:156]: If only we could make this callout appear flashing in 1000pt font.

With the model that's currently in use, it seems like there's a wider variation in the results we're getting (mine has jumped steps in a few places compared to the text), so we'll really need to set folks expectations to read this as one possible flow through the experience, and that there's is likely to differ significantly.

File: working-w-copilot/copilot-in-projects-refactoring.md [hunk 10]
```diff
@@ -115,240 +145,165 @@ Remember, the tests generated during the Copilot run in this lesson may not matc
...
   ```
 </details>
 
 #### Testing `update_book`
 
-With `validate_book` we were lucky, and Copilot outlined all of the scenarios that we intended to test, but that will not always be the case. We won't go step by step again, but let's apply the same approach we took with `validate_model` to testing our remaining dependencies `update_book` and `delete_book`.
+With `validate_book` we were lucky, and Copilot outlined all of the scenarios that we intended to test, but that will not always be the case. We won't go step by step for each dependency in this lesson, but let's apply the same approach we took with `validate_model` to testing our remaining dependencies `update_book` and `delete_book`.
 
-We'll start by using Copilot inline chat to ask for help generating tests for the `update_book` function. Here's what Copilot suggested:
+We'll start by using Copilot inline chat and the `/tests` shortcut to ask for help generating tests for the `update_book` function. Here's what Copilot suggested:
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-refactoring.md]: Can we refer to `update_book` as an endpoint? Calling it just a function here souds like we're talking about another helper, like validate_model was.
```suggestion
We'll start by using Copilot inline chat and the `/tests` shortcut to ask for help generating tests for the `update_book` endpoint. Here's what Copilot suggested:
```

File: working-w-copilot/copilot-in-projects-refactoring.md [hunk 11]
```diff
@@ -115,240 +145,165 @@ Remember, the tests generated during the Copilot run in this lesson may not matc
...
-- return an error if a book with the id doesn't exist
+We are using the database now, but from these new changes we need to address that:
+- we must import the `Book` class
+- the test is using a fixture `db_session` we haven't created
+- Copilot chose deprecated querying syntax `db_session.query(Book).get(1)`. 
+
+The overall frame of the tests is looking good so we can press "Accept" and save the file. We could see about writing a fixture `db_session` like Copilot suggested, but we will choose to import `db` into the test file since we need to use both the `select` method and `session` attribute from `db` to query for a specific record. 
+
+If we import `Book` and `db` and change our querying syntax over to using `db.select` and `db.session` our invalid and deprecated syntax issues are solved, but if we try to run the tests, we'll see a `JSONDecodeError`. This is because the line `response.get_json()` will raise a `JSONDecodeError` when the `response` object has an empty body. We'll address this last issue by either manually or with Copilot's help replacing the `response_body` assert with one that checks if `response.content_length is None`.
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-refactoring.md]: I think these commas help read more smoothly.
```suggestion
If we import `Book` and `db` and change our querying syntax over to using `db.select` and `db.session`, our invalid and deprecated syntax issues are solved, but if we try to run the tests, we'll see a `JSONDecodeError`. This is because the line `response.get_json()` will raise a `JSONDecodeError` when the `response` object has an empty body. We'll address this last issue by, either manually or with Copilot's help, replacing the `response_body` assert with one that checks if `response.content_length is None`. 
```

File: working-w-copilot/copilot-in-projects-refactoring.md [hunk 12]
```diff
@@ -115,240 +145,165 @@ Remember, the tests generated during the Copilot run in this lesson may not matc
...
+    assert book.title == "Ocean Book"
+    assert book.description == "watr 4evr"
+
+  def test_validate_book_missing_record(two_saved_books):
+    # Act & Assert
+    with pytest.raises(HTTPException) as error:
+        validate_book(3)
+
+    response = error.value.response
+    assert response.status == "404 NOT FOUND"
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-refactoring.md:200]: I never noticed the status lines were so shouty.

File: working-w-copilot/copilot-in-projects-refactoring.md [hunk 13]
```diff
@@ -358,161 +313,153 @@ However, there are other scenarios that are worth testing. What happens if the r
...
+The assertions for these tests assume that a 400 status code will be sent in case of an error, and also made an assumption about what the error message in the response would be. If we navigate to the `update_book` function in `book_routes.py`, we see that there actually isn't any error handling for this scenario. These tests fail because our application crashes with a `KeyError` when required data is missing from the request dictionary.
 
 While it's possible that Copilot could have written tests that expected a crash (which would look like a `500 Internal Server Error`), it's much more common that these scenarios would result in a `400 Bad Request`. So even though Copilot generated failing tests, it turns out that this helped us by highlighting error handling that would be useful to have. Rather than changing the tests to expect a crash, we're going to add error handling to `update_book` to make our app more robust, then update our test assertions to check for the correct message.
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-refactoring.md:429]: Is this something that we should also plan to move earlier into the relevant lesson that has the PUT endpoint, similar to what we need to do to straighten out the POST situation?

**Review by anselrognlie — APPROVED** [review]
LGTM. There are just a couple of small edit suggestions. Most of the things I posted about were places my run through differed from the updates. I think we're just going to really stress with student how variable their experience is going to be.

Also wondering whether the PUT endpoint error handling should move earlier in the flow (similar to what we need to do for the POST in order to untangle the to_dict/from_dict refactor ordering).

File: working-w-copilot/copilot-in-projects-new-code.md (hunk 1 above)
Comment (kelsey-steven-ada) [reply working-w-copilot/copilot-in-projects-new-code.md:60]: Huh, I didn't change any options and it always opened to the side when I was running through it locally the first time. Doing from the coffee shop on my single screen it is opening the inline view like you're seeing.

File: working-w-copilot/copilot-in-projects-new-code.md (hunk 3 above)
Comment (kelsey-steven-ada) [reply working-w-copilot/copilot-in-projects-new-code.md]: Sounds good, I can add a few sentences on saving the file in the proper structure to hopefully help folks no get caught up on it.

File: working-w-copilot/copilot-in-projects-refactoring.md (hunk 9 above)
Comment (kelsey-steven-ada) [reply working-w-copilot/copilot-in-projects-refactoring.md:156]: I can add a reminder at the outset of each lesson that we are working with a generative tool and what they see while following the same steps may vary greatly from the results in the example we are walking through in the lesson.

File: working-w-copilot/copilot-in-projects-refactoring.md (hunk 13 above)
Comment (kelsey-steven-ada) [reply working-w-copilot/copilot-in-projects-refactoring.md:429]: Yeah, there's no reason we shouldn't build in the error handling when creating the PUT route and would mean we likely wouldn't hit this hiccup here.

File: working-w-copilot/copilot-in-projects-new-code.md (hunk 1 above)
Comment (kelsey-steven-ada) [reply working-w-copilot/copilot-in-projects-new-code.md:60]: I took a screenshot when I ran though this again, do you think we should show both?

Comment (anselrognlie) [reply working-w-copilot/copilot-in-projects-new-code.md:60]: I can't predict what anyone is going to see. Could be something entirely different. Maybe this just falls under the banner of "things the reader should be ready to adjust to?"

File: working-w-copilot/copilot-in-projects-new-code.md (hunk 3 above)
Comment (anselrognlie) [reply working-w-copilot/copilot-in-projects-new-code.md]: Since we won't be sticking around in this branch beyond this lesson, leaving it where it is might even be fine as long as we have that little nudge. But as a new reader, I don't know what the longevity of this code is going to be, so I think that's where my inclination to want to put it in the "right" place was coming from. So if we don't want fols to get caught up here, maybe even just saying that it's OK to leave it in the same folder since we're just using this as a temporary example could be enough. So whichever makes the most sense to you seems fine.

File: working-w-copilot/copilot-in-projects-refactoring.md (hunk 13 above)
Comment (anselrognlie) [reply working-w-copilot/copilot-in-projects-refactoring.md:429]: OK. I did mention this in the Asana ticket I linked to earlier.

File: working-w-copilot/copilot-in-projects-new-code.md [hunk 14]
```diff
@@ -8,6 +8,17 @@ Our goals for this lesson are to show how we can use Copilot to:
 - quickly write new code that follows a consistent format
 - help us with generating test cases
 
+### !callout-info
+
+## Generative tools will give you different responses
+
+As you work through this lesson, you will likely get different results from the prompts you submit than what we show through the lesson. This is expected!
+
+We are working with a generative AI tool and they are not guaranteed to return the same or even a similar response for the same input. Part of adjusting to working with AI tools is getting comfortable with the variability of their responses and then fine tuning our prompts, regenerating responses, and manually updating generated code until we have something that meets our needs.
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-new-code.md]: Adds line breaks and a ntoe about the UI in the extension.
```suggestion
As you work through this lesson, you will likely get different results from the prompts you submit than what we show through the lesson. This is expected!

<br>

We are working with a generative AI tool and they are not guaranteed to return the same or even a similar response for the same input. Part of adjusting to working with AI tools is getting comfortable with the variability of their responses and then fine tuning our prompts, regenerating responses, and manually updating generated code until we have something that meets our needs.

<br>

Even the Copilot extension itself is updated regularly, so the way the UI looks or where certain features are located may differ from that shown in this lesson. So if a screenshot looks a little different from what you see in your own VS Code, try to find the equivalent feature in your version of the extension.
```

File: working-w-copilot/copilot-in-projects-refactoring.md [hunk 15]
```diff
@@ -8,17 +8,27 @@ Our goals for this lesson are to:
 - identify where Copilot can assist us with refactoring
 - get more experience adding and updating test cases with Copilot
 
+### !callout-info
+
+## Generative tools will give you different responses
+
+As you work through this lesson, you will likely get different results from the prompts you submit than what we show through the lesson. This is expected!
+
+We are working with a generative AI tool and they are not guaranteed to return the same or even a similar response for the same input. Part of adjusting to working with AI tools is getting comfortable with the variability of their responses and then fine tuning our prompts, regenerating responses, and manually updating generated code until we have something that meets our needs.
```
Comment (anselrognlie) [inline working-w-copilot/copilot-in-projects-refactoring.md]: Adds line breaks and a ntoe about the UI in the extension.
```suggestion
As you work through this lesson, you will likely get different results from the prompts you submit than what we show through the lesson. This is expected!

<br>

We are working with a generative AI tool and they are not guaranteed to return the same or even a similar response for the same input. Part of adjusting to working with AI tools is getting comfortable with the variability of their responses and then fine tuning our prompts, regenerating responses, and manually updating generated code until we have something that meets our needs.

<br>

Even the Copilot extension itself is updated regularly, so the way the UI looks or where certain features are located may differ from that shown in this lesson. So if a screenshot looks a little different from what you see in your own VS Code, try to find the equivalent feature in your version of the extension.
```

**Review by anselrognlie — APPROVED** [review]
Added a suggestion to the newly added "take care" callouts.
